// pixijs/compressed-textures/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _const = require('./const.js');
require('./resources/index.js');
require('./loaders/index.js');
require('./parsers/index.js');
var BlobResource = require('./resources/BlobResource.js');
var CompressedTextureResource = require('./resources/CompressedTextureResource.js');
var detectCompressedTextures = require('./loaders/detectCompressedTextures.js');
var loadDDS = require('./loaders/loadDDS.js');
var loadKTX = require('./loaders/loadKTX.js');
var resolveCompressedTextureUrl = require('./loaders/resolveCompressedTextureUrl.js');
var parseDDS = require('./parsers/parseDDS.js');
var parseKTX = require('./parsers/parseKTX.js');



exports.INTERNAL_FORMATS = _const.INTERNAL_FORMATS;
exports.INTERNAL_FORMAT_TO_BYTES_PER_PIXEL = _const.INTERNAL_FORMAT_TO_BYTES_PER_PIXEL;
exports.BlobResource = BlobResource.BlobResource;
exports.CompressedTextureResource = CompressedTextureResource.CompressedTextureResource;
exports.detectCompressedTextures = detectCompressedTextures.detectCompressedTextures;
exports.loadDDS = loadDDS.loadDDS;
exports.loadKTX = loadKTX.loadKTX;
exports.resolveCompressedTextureUrl = resolveCompressedTextureUrl.resolveCompressedTextureUrl;
exports.parseDDS = parseDDS.parseDDS;
exports.FORMATS_TO_COMPONENTS = parseKTX.FORMATS_TO_COMPONENTS;
exports.TYPES_TO_BYTES_PER_COMPONENT = parseKTX.TYPES_TO_BYTES_PER_COMPONENT;
exports.TYPES_TO_BYTES_PER_PIXEL = parseKTX.TYPES_TO_BYTES_PER_PIXEL;
exports.parseKTX = parseKTX.parseKTX;
