// pixijs/utils/index.mjs
export { isMobile } from 'pixijs/settings';
export { default as EventEmitter } from 'eventemitter3';
export { default as earcut } from 'earcut';
export { url } from './url.mjs';
export { path } from './path.mjs';
import './settings.mjs';
export { sayHello, skipHello } from './browser/hello.mjs';
export { isWebGLSupported } from './browser/isWebGLSupported.mjs';
export { hex2rgb, hex2string, rgb2hex, string2hex } from './color/hex.mjs';
export { correctBlendMode, premultiplyBlendMode, premultiplyRgba, premultiplyTint, premultiplyTintToRgba } from './color/premultiply.mjs';
export { createIndicesForQuads } from './data/createIndicesForQuads.mjs';
export { getBufferType } from './data/getBufferType.mjs';
export { interleaveTypedArrays } from './data/interleaveTypedArrays.mjs';
export { isPow2, log2, nextPow2 } from './data/pow2.mjs';
export { removeItems } from './data/removeItems.mjs';
export { sign } from './data/sign.mjs';
export { uid } from './data/uid.mjs';
export { deprecation } from './logging/deprecation.mjs';
export { BaseTextureCache, ProgramCache, TextureCache, clearTextureCache, destroyTextureCache } from './media/caches.mjs';
export { CanvasRenderTarget } from './media/CanvasRenderTarget.mjs';
export { trimCanvas } from './media/trimCanvas.mjs';
export { decomposeDataUri } from './network/decomposeDataUri.mjs';
export { determineCrossOrigin } from './network/determineCrossOrigin.mjs';
export { getResolutionOfUrl } from './network/getResolutionOfUrl.mjs';
export { DATA_URI } from './const.mjs';
import './types/index.mjs';
