// pixijs/basis/Basis.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('pixijs/core');
var compressedTextures = require('pixijs/compressed-textures');

var BASIS_FORMATS = /* @__PURE__ */ ((BASIS_FORMATS2) => {
  BASIS_FORMATS2[BASIS_FORMATS2["cTFETC1"] = 0] = "cTFETC1";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFETC2"] = 1] = "cTFETC2";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBC1"] = 2] = "cTFBC1";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBC3"] = 3] = "cTFBC3";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBC4"] = 4] = "cTFBC4";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBC5"] = 5] = "cTFBC5";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBC7"] = 6] = "cTFBC7";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFPVRTC1_4_RGB"] = 8] = "cTFPVRTC1_4_RGB";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFPVRTC1_4_RGBA"] = 9] = "cTFPVRTC1_4_RGBA";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFASTC_4x4"] = 10] = "cTFASTC_4x4";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFATC_RGB"] = 11] = "cTFATC_RGB";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFATC_RGBA_INTERPOLATED_ALPHA"] = 12] = "cTFATC_RGBA_INTERPOLATED_ALPHA";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFRGBA32"] = 13] = "cTFRGBA32";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFRGB565"] = 14] = "cTFRGB565";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFBGR565"] = 15] = "cTFBGR565";
  BASIS_FORMATS2[BASIS_FORMATS2["cTFRGBA4444"] = 16] = "cTFRGBA4444";
  return BASIS_FORMATS2;
})(BASIS_FORMATS || {});
const BASIS_FORMAT_TO_INTERNAL_FORMAT = {
  [0 /* cTFETC1 */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGB_ETC1_WEBGL,
  [2 /* cTFBC1 */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGB_S3TC_DXT1_EXT,
  [3 /* cTFBC3 */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGBA_S3TC_DXT5_EXT,
  [8 /* cTFPVRTC1_4_RGB */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGB_PVRTC_4BPPV1_IMG,
  [9 /* cTFPVRTC1_4_RGBA */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
  [11 /* cTFATC_RGB */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGB_ATC_WEBGL,
  [10 /* cTFASTC_4x4 */]: compressedTextures.INTERNAL_FORMATS.COMPRESSED_RGBA_ASTC_4x4_KHR
};
const BASIS_FORMAT_TO_TYPE = {
  [13 /* cTFRGBA32 */]: core.TYPES.UNSIGNED_BYTE,
  [14 /* cTFRGB565 */]: core.TYPES.UNSIGNED_SHORT_5_6_5,
  [16 /* cTFRGBA4444 */]: core.TYPES.UNSIGNED_SHORT_4_4_4_4
};
const INTERNAL_FORMAT_TO_BASIS_FORMAT = Object.keys(BASIS_FORMAT_TO_INTERNAL_FORMAT).map((key) => Number(key)).reduce((reverseMap, basisFormat) => {
  reverseMap[BASIS_FORMAT_TO_INTERNAL_FORMAT[basisFormat]] = basisFormat;
  return reverseMap;
}, {});
const BASIS_FORMATS_ALPHA = {
  [3 /* cTFBC3 */]: true,
  [9 /* cTFPVRTC1_4_RGBA */]: true,
  [10 /* cTFASTC_4x4 */]: true
};

exports.BASIS_FORMATS = BASIS_FORMATS;
exports.BASIS_FORMATS_ALPHA = BASIS_FORMATS_ALPHA;
exports.BASIS_FORMAT_TO_INTERNAL_FORMAT = BASIS_FORMAT_TO_INTERNAL_FORMAT;
exports.BASIS_FORMAT_TO_TYPE = BASIS_FORMAT_TO_TYPE;
exports.INTERNAL_FORMAT_TO_BASIS_FORMAT = INTERNAL_FORMAT_TO_BASIS_FORMAT;
