// pixijs/core/shader/Program.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./utils/index.js');
var utils = require('pixijs/utils');
var defaultProgram$1 = require('./defaultProgram.js');
var defaultProgram = require('./defaultProgram2.js');
var settings = require('pixijs/settings');
var constants = require('pixijs/constants');
var setPrecision = require('./utils/setPrecision.js');
var getMaxFragmentPrecision = require('./utils/getMaxFragmentPrecision.js');

let UID = 0;
const nameCache = {};
class Program {
  constructor(vertexSrc, fragmentSrc, name = "pixi-shader", extra = {}) {
    this.extra = {};
    this.id = UID++;
    this.vertexSrc = vertexSrc || Program.defaultVertexSrc;
    this.fragmentSrc = fragmentSrc || Program.defaultFragmentSrc;
    this.vertexSrc = this.vertexSrc.trim();
    this.fragmentSrc = this.fragmentSrc.trim();
    this.extra = extra;
    if (this.vertexSrc.substring(0, 8) !== "#version") {
      name = name.replace(/\s+/g, "-");
      if (nameCache[name]) {
        nameCache[name]++;
        name += `-${nameCache[name]}`;
      } else {
        nameCache[name] = 1;
      }
      this.vertexSrc = `#define SHADER_NAME ${name}
${this.vertexSrc}`;
      this.fragmentSrc = `#define SHADER_NAME ${name}
${this.fragmentSrc}`;
      this.vertexSrc = setPrecision.setPrecision(this.vertexSrc, settings.settings.PRECISION_VERTEX, constants.PRECISION.HIGH);
      this.fragmentSrc = setPrecision.setPrecision(this.fragmentSrc, settings.settings.PRECISION_FRAGMENT, getMaxFragmentPrecision.getMaxFragmentPrecision());
    }
    this.glPrograms = {};
    this.syncUniforms = null;
  }
  static get defaultVertexSrc() {
    return defaultProgram["default"];
  }
  static get defaultFragmentSrc() {
    return defaultProgram$1["default"];
  }
  static from(vertexSrc, fragmentSrc, name) {
    const key = vertexSrc + fragmentSrc;
    let program = utils.ProgramCache[key];
    if (!program) {
      utils.ProgramCache[key] = program = new Program(vertexSrc, fragmentSrc, name);
    }
    return program;
  }
}

exports.Program = Program;
