// pixijs/node/adapter/loadNodeBase64.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var canvasModule = require('canvas');
var core = require('pixijs/core');
var NodeCanvasElement = require('./NodeCanvasElement.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var canvasModule__default = /*#__PURE__*/_interopDefaultLegacy(canvasModule);

const { loadImage } = canvasModule__default["default"];
const validMimes = ["image/png", "image/jpg", "image/jpeg", "image/svg"];
function isSupportedDataURL(url) {
  const match = url.match(/^data:([^;]+);base64,/);
  if (!match)
    return false;
  const mimeType = match[1];
  return validMimes.includes(mimeType);
}
const loadNodeBase64 = {
  extension: core.ExtensionType.LoadParser,
  test(url) {
    return isSupportedDataURL(url);
  },
  async load(url, asset) {
    const image = await loadImage(url);
    const canvas = new NodeCanvasElement.NodeCanvasElement(image.width, image.height);
    const ctx = canvas.getContext("2d");
    ctx.drawImage(image, 0, 0);
    const texture = core.Texture.from(canvas, {
      resolution: core.utils.getResolutionOfUrl(url),
      ...asset.data
    });
    return texture;
  },
  unload(texture) {
    texture.destroy(true);
  }
};
core.extensions.add(loadNodeBase64);

exports.loadNodeBase64 = loadNodeBase64;
