// pixijs/node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./adapter/index.js');
var core = require('pixijs/core');
var filterAlpha = require('pixijs/filter/alpha');
var filterBlur = require('pixijs/filter/blur');
var filterColorMatrix = require('pixijs/filter/color-matrix');
var filterDisplacement = require('pixijs/filter/displacement');
var filterFxaa = require('pixijs/filter/fxaa');
var filterNoise = require('pixijs/filter/noise');
require('pixijs/display/cache-as-bitmap');
require('pixijs/display/get-child-by-name');
require('pixijs/display/get-global-position');
var assets = require('pixijs/assets');
var app = require('pixijs/app');
var textBitmap = require('pixijs/text-bitmap');
var display = require('pixijs/display');
var extract = require('pixijs/extract');
var graphics = require('pixijs/graphics');
var mesh = require('pixijs/mesh');
var meshExtras = require('pixijs/mesh/extras');
var particleContainer = require('pixijs/particle-container');
var prepare = require('pixijs/prepare');
var sprite = require('pixijs/sprite');
var spriteAnimated = require('pixijs/sprite-animated');
var spriteTiling = require('pixijs/sprite-tiling');
var spritesheet = require('pixijs/spritesheet');
var text = require('pixijs/text');
var NodeCanvasResource = require('./adapter/NodeCanvasResource.js');
var NodeCanvasElement = require('./adapter/NodeCanvasElement.js');
var loadNodeBitmapFont = require('./adapter/loadNodeBitmapFont.js');
var loadNodeFont = require('./adapter/loadNodeFont.js');
var loadNodeTexture = require('./adapter/loadNodeTexture.js');
var loadNodeBase64 = require('./adapter/loadNodeBase64.js');
var adapter = require('./adapter/adapter.js');

core.extensions.remove(assets.loadTextures, assets.loadWebFont, textBitmap.loadBitmapFont, app.ResizePlugin);
core.INSTALLED.length = 0;
core.INSTALLED.push(NodeCanvasResource.NodeCanvasResource);
const filters = {
  AlphaFilter: filterAlpha.AlphaFilter,
  BlurFilter: filterBlur.BlurFilter,
  BlurFilterPass: filterBlur.BlurFilterPass,
  ColorMatrixFilter: filterColorMatrix.ColorMatrixFilter,
  DisplacementFilter: filterDisplacement.DisplacementFilter,
  FXAAFilter: filterFxaa.FXAAFilter,
  NoiseFilter: filterNoise.NoiseFilter
};

Object.defineProperty(exports, 'settings', {
    enumerable: true,
    get: function () { return core.settings; }
});
exports.NodeCanvasResource = NodeCanvasResource.NodeCanvasResource;
exports.NodeCanvasElement = NodeCanvasElement.NodeCanvasElement;
exports.loadNodeBitmapFont = loadNodeBitmapFont.loadNodeBitmapFont;
exports.loadNodeFont = loadNodeFont.loadNodeFont;
exports.loadNodeTexture = loadNodeTexture.loadNodeTexture;
exports.loadNodeBase64 = loadNodeBase64.loadNodeBase64;
exports.NodeAdapter = adapter.NodeAdapter;
exports.filters = filters;
Object.keys(core).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return core[k]; }
    });
});
Object.keys(assets).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return assets[k]; }
    });
});
Object.keys(app).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return app[k]; }
    });
});
Object.keys(textBitmap).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return textBitmap[k]; }
    });
});
Object.keys(display).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return display[k]; }
    });
});
Object.keys(extract).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return extract[k]; }
    });
});
Object.keys(graphics).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return graphics[k]; }
    });
});
Object.keys(mesh).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return mesh[k]; }
    });
});
Object.keys(meshExtras).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return meshExtras[k]; }
    });
});
Object.keys(particleContainer).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return particleContainer[k]; }
    });
});
Object.keys(prepare).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return prepare[k]; }
    });
});
Object.keys(sprite).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return sprite[k]; }
    });
});
Object.keys(spriteAnimated).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return spriteAnimated[k]; }
    });
});
Object.keys(spriteTiling).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return spriteTiling[k]; }
    });
});
Object.keys(spritesheet).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return spritesheet[k]; }
    });
});
Object.keys(text).forEach(function (k) {
    if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
        enumerable: true,
        get: function () { return text[k]; }
    });
});
