// pixijs/node/index.mjs
import './adapter/index.mjs';
import { extensions, INSTALLED } from 'pixijs/core';
export * from 'pixijs/core';
export { settings } from 'pixijs/core';
import { AlphaFilter } from 'pixijs/filter/alpha';
import { BlurFilter, BlurFilterPass } from 'pixijs/filter/blur';
import { ColorMatrixFilter } from 'pixijs/filter/color-matrix';
import { DisplacementFilter } from 'pixijs/filter/displacement';
import { FXAAFilter } from 'pixijs/filter/fxaa';
import { NoiseFilter } from 'pixijs/filter/noise';
import 'pixijs/display/cache-as-bitmap';
import 'pixijs/display/get-child-by-name';
import 'pixijs/display/get-global-position';
import { loadTextures, loadWebFont } from 'pixijs/assets';
export * from 'pixijs/assets';
import { ResizePlugin } from 'pixijs/app';
export * from 'pixijs/app';
import { loadBitmapFont } from 'pixijs/text-bitmap';
export * from 'pixijs/text-bitmap';
export * from 'pixijs/display';
export * from 'pixijs/extract';
export * from 'pixijs/graphics';
export * from 'pixijs/mesh';
export * from 'pixijs/mesh/extras';
export * from 'pixijs/particle-container';
export * from 'pixijs/prepare';
export * from 'pixijs/sprite';
export * from 'pixijs/sprite-animated';
export * from 'pixijs/sprite-tiling';
export * from 'pixijs/spritesheet';
export * from 'pixijs/text';
import { NodeCanvasResource } from './adapter/NodeCanvasResource.mjs';
export { NodeCanvasResource } from './adapter/NodeCanvasResource.mjs';
export { NodeCanvasElement } from './adapter/NodeCanvasElement.mjs';
export { loadNodeBitmapFont } from './adapter/loadNodeBitmapFont.mjs';
export { loadNodeFont } from './adapter/loadNodeFont.mjs';
export { loadNodeTexture } from './adapter/loadNodeTexture.mjs';
export { loadNodeBase64 } from './adapter/loadNodeBase64.mjs';
export { NodeAdapter } from './adapter/adapter.mjs';

extensions.remove(loadTextures, loadWebFont, loadBitmapFont, ResizePlugin);
INSTALLED.length = 0;
INSTALLED.push(NodeCanvasResource);
const filters = {
  AlphaFilter,
  BlurFilter,
  BlurFilterPass,
  ColorMatrixFilter,
  DisplacementFilter,
  FXAAFilter,
  NoiseFilter
};

export { filters };
